File: app/vectorbt_test/core/factors.py
from vectorbt_test.core.nodes import Node, FeatureNode, NodeType, NodeDType
from vectorbt_test.core.node_builder import NodeBuilder
from vectorbt_test.core.functions import Rank
from vectorbt_test.core.context import PortfolioContext
from vectorbt_test.core.registry import NodeRegistry, NodeMeta, NodeParam
from vectorbt_test.core.base import Scope
import pandas as pd


class Factor(FeatureNode):
    scope = Scope.TS
    dtype = NodeDType.Numeric

    def __init__(self, name: str, expr_str: str | Node):
        super().__init__(NodeType.Factor)
        self._name = name
        if isinstance(expr_str, str):
            self.node = NodeBuilder().build(expr_str)
        else:
            self.node = expr_str

        assert self.type == NodeType.Factor
        
    @property
    def name(self):
        return self._name
    
    def _args(self):
        return [self.name, self.node.cache_key()] + super()._args()
    
    def compute(self, data: pd.DataFrame, context: PortfolioContext):
        return self.node.evaluate(data, context)

    def score(self, data: pd.DataFrame, context: PortfolioContext):
        s = self.evaluate(data, context)
        return context.data_adapter.cs_zscore(s)

    def rank(self):
        return Rank(self)


class BoolFactor(Factor):
    dtype = NodeDType.Bool


class SignalToFactor(Factor):
    def __init__(self, signal_node):
        if signal_node.dtype != NodeDType.Signal:
            raise TypeError("Expect signal")
        super().__init__("signal_node", signal_node)

    def compute(self, data, context):
        return self.node.evaluate(data, context).astype(float)


class FactorWrapper(Factor):
    def __init__(self, node: Node):
        if node.dtype != NodeDType.Numeric:
            raise TypeError("Expect numberic type of node")
        super().__init__(node.__class__.__name__, node)

    def compute(self, data, context):
        return self.node.evaluate(data, context)


def wrap_numberic_node_as_factor(node: Node):
    if node.type == NodeType.Signal and node.dtype == NodeDType.Numeric:
        return SignalToFactor(node)
    return FactorWrapper(node)


class GeneralFactor(Factor):
    def __init__(self, name: str, expr_str: str):
        super().__init__(name, expr_str)


def register_factors():
    NodeRegistry.register(
        "GFactor",
        lambda name, expr_str: GeneralFactor(name, expr_str),
        NodeMeta(
            name="GFactor",
            group="factor",
            desc="GeneralFactor to create general factors",
            params=[
                NodeParam("name", "str", None, "Factor name"),
                NodeParam("expr_str", "str", None, "Expression string")
            ]
        ))

# class Rank(FeatureNode):
#     scope = Scope.CS
#     def __init__(self, node):
#         super().__init__(node.type)
#         self.node = node

#     def compute(self, data, context: PortfolioContext):
#         x = self.node.evaluate(data, context)
#         return x.rank(axis=1, pct=True)

    
# NodeRegistry.register(
#     "momentum_score",
#     lambda: DBNode("momentum_score"),
#     NodeMeta(
#         name="momentum_score",
#         group="factor",
#         desc="动量因子"
#     )
# )
# NodeRegistry.register(
#     "close",
#     lambda: DBNode("close"),
#     NodeMeta(
#         name="close",
#         group="raw",
#         desc="收盘价"
#     )
# )

# if __name__ == "__main__":
#     from vectorbt_test.core.expr_parser import NodeBuilder

#     builder = NodeBuilder()

#     expr = builder.build("0.7*ma5 + 0.3*macd")

#     factor = FactorNode("trend", expr)

#     context = {
#         "db_df": df   # 从数据库读取的指标
#     }

#     score = factor.score(df, context)

File: quant-mantis/src/model/dsl_node/factors.ts
import { NodeType, NodeDType, Scope, PortfolioContext } from "./types";
import { Node, FeatureNode } from "./nodes";

// 因子基类
export class Factor extends FeatureNode {
    node: Node;


    constructor(name: string | null, exprStr: string | Node) {
        super(NodeType.Factor);
        this.dtype = NodeDType.NUMERIC;
        this.scope = Scope.TS
        this._name = name || this.constructor.name;

        // 忽略 NodeBuilder 构建逻辑，直接赋值
        this.node = typeof exprStr === "string" ? {} as Node : exprStr;

        if (this.type !== NodeType.Factor) {
            throw new Error("Node type must be Factor");
        }
    }

    get name(): string {
        return this._name;
    }

    _args(): any[] {
        return [this.name, this.node.cacheKey(), ...super._args()];
    }

    // 忽略 compute 业务逻辑
    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }

    // 忽略 score 业务逻辑
    score(_data: any, _context: PortfolioContext): any {
        return null;
    }

    rank(): Rank {
        return new Rank(this);
    }
}

// 布尔因子
export class BoolFactor extends Factor {
    constructor(name: string | null, exprStr: string | Node) {
        super(name, exprStr);
        this.dtype = NodeDType.BOOL;
    }
}

// 信号转因子
export class SignalToFactor extends Factor {
    constructor(name: string | null, signalNode: Node) {
        if (signalNode.dtype !== NodeDType.SIGNAL) {
            throw new TypeError("Expect signal");
        }
        super(name, signalNode);
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 因子包装器
export class FactorWrapper extends Factor {
    constructor(name: string | null, node: Node) {
        if (node.dtype !== NodeDType.NUMERIC) {
            throw new TypeError("Expect numeric type of node");
        }
        super(name, node);
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 工具函数
export function wrapNumericNodeAsFactor(name: string | null, node: Node): Factor {
    if (node.type === NodeType.Signal && node.dtype === NodeDType.NUMERIC) {
        return new SignalToFactor(name, node);
    }
    return new FactorWrapper(name, node);
}

// 通用因子
export class GeneralFactor extends Factor {
    constructor(name: string, exprStr: string) {
        super(name, exprStr);
    }
}

// 排名类（移至此处，对应 Python 中 functions.py 的 Rank）
export class Rank extends FeatureNode {
    node: Node;
    scope = Scope.CS

    constructor(node: Node) {
        super(NodeType.Factor);
        this.dtype = NodeDType.NUMERIC;

        if (node.dtype !== NodeDType.NUMERIC) {
            throw new TypeError("rank() requires numeric input");
        }
        this.node = node;
    }

    _args(): any[] {
        return [this.node.cacheKey(), ...super._args()];
    }

    compute(_data: any, _context: PortfolioContext): any {
        return null;
    }
}

// 注册函数（简化）
export function registerFactors(): void {
    // 忽略注册逻辑，仅保留函数结构
}